docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,147 @@ Every inter-canister call is guaranteed to receive a response, either from the c
+
+#### Recovery
+
+The journal ensures the application knows that recovery from an error is needed, and aids in making recovery decisions. In order to support the recovery process, the journal should support querying all unresolved tasks of a certain type, and tasks of a certain type that resulted in an error. Given an intent, the journal should also be able to return the result if it exists and indicate if it does not exist.
+
+Extending the ledger example above, the recovery process can look like the following
+
+1. There is a panic and the status of the ledger call is unknown. However, the journal has recorded that a call to transfer with particular parameters and memo has been made, and the deduplication timestamp of the transfer. 
+1. The app calls the ledger to determine whether a transaction with the journaled parameters has succeeded on the ledger. Note that with the ICP ledger, there is a small possibility the transaction is stuck in the message queue, so one needs to wait 24 hours from the deduplication time stamp to ensure the message has either succeeded or failed and is not in transit. Therefore if the ledger does not contain the transaction and it is still within 24 hours of the deduplication timestamp, this step must be repeated.
+1. The app journals the result of the transfer call.
+1. The app journals the intention to update internal state according to the result of the transfer call, then updates the internal state, and finally journals the result of the attempt to update the internal state. (Journaling this step is still useful even if it does not contain outcalls, because outcalls may be introduced later, and the step could conflict with other processes that are not atomic.)
+
+Note that recovery can be complex in many cases. Therefore, it may be expedient to rely on manual recovery, then progressively implement automated recovery.
+
+#### Example Journaling Structures
+
+The journaling data structures should support the following queries.
+- What are the unresolved actions for a particular user and action type?
+- What are the unresolved actions that have been open for longer than some time T?
+
+Map: Journal by caller
+- Key: (caller, action type) – e.g. (Alice, Withdraw)
+- Value: Transaction journal struct
+
+Map: Transactions by timestamp (ordered map, could use an additional list for the ordering in Motoko)
+- Key: timestamp 
+- Value: list of references to Transaction journal structs
+
+Transaction journal struct
+- Meta data
+  - transaction, including memo
+  - timestamp
+- List of events, e.g. initialized, returned/success, returned/error with error condition
+- Number of times attempted and failed to progress
+
+#### Example Journaling Flows
```

Comment:
@andrew-lee-work Can we also link to implementations of this journaling?

Reply:
There is an implementation of journaling in https://github.com/GoldDAO/gldt-swap/blob/develop/canister/gldt_core/src/lib.rs#L654. I will add it.

Reply:
Please use a permalink so that it stays stable; the link is already broken for me.

Reply:
OK, changing to https://github.com/GoldDAO/gldt-swap/blob/ledger-v1.0.0/canister/gldt_core/src/lib.rs . This is a tag, so is more permanent.